### ISSUE TITLE
[mlir][shard] scatter/all_to_all/shard_slice verifiers index the shape with an unchecked axis attribute

### ISSUE THREAD

**Issue by alepot55** [llvm/llvm-project#218212]
`shard.gather` bounds-checks its axis attribute before using it. Its siblings do not, and index the shape with an unvalidated value.

`verifyGatherOperandAndResultShape` (ShardOps.cpp:1056) starts with:

```cpp
if (gatherAxis < 0 || gatherAxis >= resultRank) {
  return emitError(result.getLoc())
         << "Gather axis " << gatherAxis << " is out of bounds [0, "
         << resultRank << ").";
}
```

`verifyScatterOrSliceOperandAndResultShape` (:1128) and `verifyAllToAllOperandAndResultShape` (:1083) have no such check and go straight to `operandType.getDimSize(tensorAxis)`. `ShapedType::getDimSize` is

```cpp
int64_t getDimSize(unsigned idx) const {
  assert(idx < getRank() && "invalid index for shaped type");
  return $_type.getShape()[idx];
}
```

so an out-of-range axis asserts in an assertions build and reads past the end of the shape otherwise. The unchecked verifier is reached from `shard.scatter` (:1243), `shard.all_to_all` (:1283), `shard.shard_slice` (:1424) and `:1455`.

Expected: a diagnostic like the one gather already emits. Actual: the verifier itself is the thing that goes out of bounds, which is an awkward place for it since the verifier is what should be catching this.

Reduced from the existing test at `mlir/test/Dialect/Shard/invalid.mlir`, changing only the axis:

```mlir
shard.grid @grid0(shape = 3)

func.func @scatter_axis_out_of_bounds(%arg0 : tensor<3xf32>) -> tensor<1xf32> {
  %0 = shard.scatter %arg0 on @grid0 grid_axes = [0]
    scatter_dim = 5 root = [1]
    : (tensor<3xf32>) -> tensor<1xf32>
  return %0 : tensor<1xf32>
}
```

I found this by reading the source and have not executed the reduced case, so please treat the repro as unconfirmed until someone runs it; the asymmetry between the three verifiers is the part I am confident about, since it is verbatim in one file.

Happy to send the patch and the `invalid.mlir` cases. Disclosure per the AI tool policy: AI assistance was used to find this and would be used for the patch; I understand the change and can defend it in review.

**Comment by alepot55** [comment]
I'm working on a patch for this.
